atomic_reactor/plugins/pre_add_image_content_manifest.py

```diff
@@ -0,0 +1,208 @@
+                                      INSPECT_ROOTFS_LAYERS,
+                                      PLUGIN_ADD_IMAGE_CONTENT_MANIFEST,
+                                      REPO_CONTENT_SETS_CONFIG)
+from atomic_reactor.plugin import PreBuildPlugin
+from atomic_reactor.plugins.pre_reactor_config import get_cachito
+from atomic_reactor.util import (base_image_is_scratch, df_parser, read_yaml,
+                                 read_yaml_from_file_path, get_retrying_requests_session,
+                                 )
+
+
+class AddImageContentManifestPlugin(PreBuildPlugin):
+    """
+    Add the ICM JSON file to the IMAGE_BUILD_INFO_DIR/content_manifests
+    directory, for the current platform. Filename will be '{IMAGE_NVR}.json'
+
+    ICM examples:
+
+    WITHOUT content_sets specified:
+
+    {
+      "metadata": {
+        "icm_version": 1,
+        "icm_spec": "https://link.to.icm.specification",
+        "image_layer_index": 3
+      },
+      "content_sets" : [],
+      "image_contents": [
+        {
+          "purl": "pkg:golang/github.com%2Frelease-engineering%2Fretrodep%2Fv2@v2.0.2",
+          "dependencies": [{"purl": "pkg:golang/github.com%2Fop%2Fgo-logging@v0.0.0"}],
+          "sources": [{"purl": "pkg:golang/github.com%2FMasterminds%2Fsemver@v1.4.2"}]
+        }
+      ]
+    }
+
+    WITH content_sets specified:
+
+    {
+      "metadata": {
+        "icm_version": 1,
+        "icm_spec": "https://link.to.icm.specification",
+        "image_layer_index": 2
+      },
+      "content_sets": [
+          "rhel-8-for-x86_64-baseos-rpms",
+          "rhel-8-for-x86_64-appstream-rpms"
+      ],
+      "image_contents": [
+        {
+          "purl": "pkg:golang/github.com%2Frelease-engineering%2Fretrodep%2Fv2@v2.0.2",
+          "dependencies": [{"purl": "pkg:golang/github.com%2Fop%2Fgo-logging@v0.0.0"}],
+          "sources": [{"purl": "pkg:golang/github.com%2FMasterminds%2Fsemver@v1.4.2"}]
+        }
+      ]
+    }
+    """
+    key = PLUGIN_ADD_IMAGE_CONTENT_MANIFEST
+    is_allowed_to_fail = False
+    minimal_icm = {
+        'metadata': {
+            'icm_version': 1,
+            'icm_spec': ('https://raw.githubusercontent.com/containerbuildsystem/atomic-reactor/'
+                         'master/atomic_reactor/schemas/content_manifest.json'),
+            'image_layer_index': -1,
+        },
+        'content_sets': [],
+        'image_contents': [],
+    }
+
+    def __init__(self, tasker, workflow, remote_source_icm_url=None, destdir=IMAGE_BUILD_INFO_DIR):
+        """
+        :param tasker: ContainerTasker instance
+        :param workflow: DockerBuildWorkflow instance
+        :param icm_url: str, URL of the ICM from the Cachito request.
+        :param destdir: image path to carry content_manifests data dir
+        """
+        super(AddImageContentManifestPlugin, self).__init__(tasker, workflow)
+        self.content_manifests_dir = os.path.join(destdir, 'content_manifests', '')
```

Comment:
What is the purpose of that extra empty string at the end?

Reply:
Oh I think I got it:
```diff
- ADD eggs-1.0-42.json /root/buildinfo/content_manifests
+ ADD eggs-1.0-42.json /root/buildinfo/content_manifests/
```
Previously, the file would be added as `content_manifests` instead of being put into the `content_manifests/` dir

Another way to achieve that would be this:
```diff
- ADD eggs-1.0-42.json /root/buildinfo/content_manifests
+ ADD eggs-1.0-42.json /root/buildinfo/content_manifests/eggs-1.0.42.json
```
IMO that is a little less confusing and more explicit, what do you think?

Reply:
@chmeliik; I'd rather keep it platform-neutral and add a comment, rather than hard-code a '/'

Reply:
I'm not saying hard-code a '/', you could change [line 193](https://github.com/ben-alkov/atomic-reactor/blob/b23f39e1a9e180a0f0ba1f655377cd38a7a4f715/atomic_reactor/plugins/pre_add_image_content_manifest.py#L193) to this instead:
```python
dest_file_name = os.path.join(self.content_manifests_dir, self.icm_file_name)
content = 'ADD {0} {1}'.format(self.icm_file_name, dest_file_name)
```

Which would give you something like the second diff in my previous comment.
IMO using the full filename is more reliable than depending on a trailing '/'

Reply:
The empty string in `os.path.join(destdir, 'content_manifests', '')` should not be necessary anymore (it will work anyway, but IMO is confusing)

Also, that `dest_file_name` should probably be `dest_file_path`, my bad